src/java.base/share/classes/java/lang/invoke/NativeMethodHandle.java

```diff
@@ -78,7 +92,9 @@ private static boolean isIllegalType(Class<?> pType) {
               || pType == float.class
               || pType == double.class
               || pType == void.class
-              || pType == Object.class);
+              || pType == Object.class
+              || pType == MemorySegment.class // Skip checking the argument preset in OpenJ9 for the moment.
+              || pType == SegmentAllocator.class); // Skip checking the argument preset OpenJ9 for the moment.
```

Comment:
> Please explain why these classes need to be allowed here. 

The leading 3 arguments of runNativeMethod() (`MemorySegment downcallAddr, SegmentAllocator segmtAllocator, 
MemorySegment stateSegmt`) at https://github.com/eclipse-openj9/openj9/blob/7bbeb0f43dcf2ac8677928512e1c134b14ae4f46/jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java#L712
```
	Object runNativeMethod(MemorySegment downcallAddr, SegmentAllocator segmtAllocator, <------
                                                  MemorySegment stateSegmt,  <------
long[] args) throws IllegalArgumentException, IllegalStateException {
...
}
```
must be allowed to ensure it passes the type check on the arguments of LambdaForm-generated bytecode in OpenJDK.

>  The comments suggest it's temporary: when will the exceptions be removed?

The prototype only supports primitives for the time being as we have not yet determined whether they must always be there for struct/union (which won't be supported until we figure out everything with primitives) from VM & JIT perspective and how to deal with this situation. So we will need to leave them in there till everything is implemented in our code and get back to check whether they need to be removed or not at that time.

Reply:
Perhaps the exceptions should only be made for parameter types and not also the return type by modifying `hasIllegalType()` instead?
```java
    private static boolean hasIllegalType(MethodType type) {
        if (isIllegalType(type.returnType()))
            return true;

        for (Class<?> pType : type.ptypes()) {
            // Specifically allow these parameter types for OpenJ9 for the moment.
            if (pType == MemorySegment.class || pType == SegmentAllocator.class) {
                continue;
            }
            if (isIllegalType(pType))
                return true;
        }

        return false;
    }
```

Reply:
That's correct for the moment but won't be necessary as ultimately the check on `MemorySegment` & `SegmentAllocator` should be moved to `isIllegalType()` (the same as the current changes) once we confirm everything in the calling path of the MH invocation from VM & JIT perspective.

Reply:
Based on your answer, we don't need to allow the return type to be `MemorySegment` or `SegmentAllocator`, so it doesn't belong in `isIllegalType()`.

Perhaps we should be even more restrictive, allowing those types only for parameters of `runNativeMethod()`?

Reply:
>  we don't need to allow the return type to be MemorySegment or SegmentAllocator, so it doesn't belong in isIllegalType().

`MemorySegment` is used for the returned `struct/union` (which we have not ye supported for now)

`SegmentAllocator` might be removed from the return type check (as it is intended to allocate the memory for `struct/union`)

Both types should be reserved for the parameter type check as they are specific to `runNativeMethod`.

> Perhaps we should be even more restrictive, allowing those types only for parameters of runNativeMethod()?

No need to worry about this given the passed-in nep (namely the bound MH) already indicates the method is `runNativeMethod`.

Overall, we don't have to emphasize the difference of the parameter & return type on the type check so as to minimize the changes to adapt them to the upstream in OpenJDK; otherwise, we have to treat them separately with different methods for the parameter & return type.

Reply:
My concern is that `hasIllegalType()` is used for other situations where we don't want those extra types to be allowed.

Reply:
There is no other case to do the type check via `hasIllegalType()` except `nativeMH.make()` (actually check  the types in `runNativeMethod` for the bound MH so as to generate the native MH) given the `nativeMH/LinkToNative` based solution is only specific to OpenJ9 rather then OpenJDK//Hotspot unless the related code are changed in the future in OpenJDK (in which case we will have to figure out other way around this situation).

Reply:
You haven't convince me that the additional types should be allowed for the return-type of the method handle being checked.

Reply:
I have no problem with the suggestion at https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/736#discussion_r1466528442 give the return type of `runNativeMethod` is `Object` (will be converted to the specific value against the actual return type) in our implementation.
```
Object runNativeMethod(MemorySegment downcallAddr, SegmentAllocator segmtAllocator, ...
```

Reply:
Updated the code against the suggestion at https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/736#discussion_r1466528442.